MobiusCore/Source/EffectHandlers/Response.swift

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2019 Spotify AB.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+
+/// A `Response` can send output and signal completion.
+/// Sending output is done with `.send` and signaling completion is done with `.end`.
+///
+/// Note: Once `.end` has been called (from any thread), the closure you provide in `onSend` will no longer be called.
+/// Note: The closure you provide in `onEnd` will only be called once when `.end` is called on this object.
+public final class Response<T> {
```

Comment:
It would be useful to harmonise the names used here with the ones used in the Kotlin experiments. Maybe that should just be a note for now, because we're not really yet aiming at creating the final API, I think. For reference, the work in progress there defines an `Effect.Callback` which fills the same role as `Response`, and has the methods `onStart(Disposable)`, `onValue` and `onComplete`. It's also to be noted that it's the same thing as a Reactive Streams [Subscriber](https://github.com/reactive-streams/reactive-streams-jvm#2-subscriber-code).

Reply:
`Response` is a subset of each of things you mentioned. I don't think it makes sense to use the same name.

Reply:
Not "a subset of each", but a subset of all, I guess you mean? I still think it makes sense to share the name, mainly because they fill exactly the same niche - enabling reporting back about the progress of handling an effect.

Reply:
I mean "of each".

`Effect.Callback` would not work since we don't have an `Effect` type. `EffectHandler.Callback` would work, but due to how Swift works we would also need to expose the underlying `Response` type, or whatever we choose to call it.

`Subscriber` would work too, but I don't think it's a very helpful name in this context.

Reply:
`EffectResponse` and `EffectCallback` would work. Would these be preferable?

Reply:
@anawara opinions?

Reply:
I agree we should use the same names. I don't really have strong opinions in which name to pick although `Callback` feels a bit more natural to me. Showing some code below to get a feeling on how it may look.

With `EffectResponse` it will look like:
```
interface Effect<T> {
    interface Response<T> {
        fun start(disposable: Disposable)
        fun send(value: T)
        fun end()
    }
    fun execute(response: Response<T>)
}
```
and when used:
```
fun <T> fromValue(value: T): Effect<T> = object : Effect<T> {
    override fun execute(response: Response<T>) = with(response) {
        start(Disposable {})
        send(value)
        end()
    }
}
```
vs `EffectCallback`
```
interface Effect<T> {
    interface Callback<T> {
        fun onStart(disposable: Disposable)
        fun onValue(value: T)
        fun onComplete()
    }
    fun execute(callback: Callback<T>)
}
```
and when used:
```
fun <T> fromValue(value: T): Effect<T> = object : Effect<T> {
    override fun execute(callback: Callback<T>) = with(callback) {
        onStart(Disposable {})
        onValue(value)
        onComplete()
    }
}
```

Reply:
I prefer `Response` because it tells you what it does as opposed to what strategy was used to implement it.

Reply:
A disadvantage of `Response` is it is usually used synchronously/for a single response, https://en.wikipedia.org/wiki/Request%E2%80%93response. This is an asynchronous use case, with multiple values possibly associated with an originating request/effect.

Reply:
I agree with Petter, the name Response indicates that it's an output that I can inspect to see a result. It indicates that the operation has already been completed and this is the wrapper type holding it. However, what we're passing in here is a continuation/callback. It's also a little odd that I pass in a response to the execute/handle function. It reads really awkwardly

Reply:
That's fair

Reply:
I'll change it to `EffectCallback`